boot/zephyr/flash_map.c

```diff
@@ -43,44 +43,59 @@ extern struct device *boot_flash_device;
 #define FLASH_MAP_ENTRY_MAGIC 0xd00dbeef
 
 struct flash_map_entry {
-    const uint32_t magic;
-    const struct flash_area area;
+    uint32_t magic;
+    struct flash_area area;
     unsigned int ref_count;
 };
 
 /*
  * The flash area describes essentially the partition table of the
  * flash.  In this case, it starts with FLASH_AREA_IMAGE_0.
  */
-static struct flash_map_entry part_map[] = {
-    {
-        .magic = FLASH_MAP_ENTRY_MAGIC,
-        .area = {
-            .fa_id = FLASH_AREA_IMAGE_0,
-            .fa_device_id = FLASH_DEVICE_ID,
-            .fa_off = FLASH_AREA_IMAGE_0_OFFSET,
-            .fa_size = FLASH_AREA_IMAGE_0_SIZE,
-        },
-    },
-    {
-        .magic = FLASH_MAP_ENTRY_MAGIC,
-        .area = {
-            .fa_id = FLASH_AREA_IMAGE_1,
-            .fa_device_id = FLASH_DEVICE_ID,
-            .fa_off = FLASH_AREA_IMAGE_1_OFFSET,
-            .fa_size = FLASH_AREA_IMAGE_1_SIZE,
-        },
-    },
-    {
-        .magic = FLASH_MAP_ENTRY_MAGIC,
-        .area = {
-            .fa_id = FLASH_AREA_IMAGE_SCRATCH,
-            .fa_device_id = FLASH_DEVICE_ID,
-            .fa_off = FLASH_AREA_IMAGE_SCRATCH_OFFSET,
-            .fa_size = FLASH_AREA_IMAGE_SCRATCH_SIZE,
-        },
+static struct flash_map_entry part_map[3] = {0};
+
+void flash_partition_map_init(uint8_t img_idx) {
+    if (img_idx == 0) {
+        part_map[0].magic = FLASH_MAP_ENTRY_MAGIC;
+        part_map[0].area.fa_id = FLASH_AREA_IMAGE_0;
+        part_map[0].area.fa_device_id = FLASH_DEVICE_ID;
+        part_map[0].area.fa_off = FLASH_AREA_IMAGE_0_OFFSET;
+        part_map[0].area.fa_size = FLASH_AREA_IMAGE_0_SIZE;
+
+        part_map[1].magic = FLASH_MAP_ENTRY_MAGIC;
+        part_map[1].area.fa_id = FLASH_AREA_IMAGE_1;
+        part_map[1].area.fa_device_id = FLASH_DEVICE_ID;
+        part_map[1].area.fa_off = FLASH_AREA_IMAGE_1_OFFSET;
+        part_map[1].area.fa_size = FLASH_AREA_IMAGE_1_SIZE;
+
+        part_map[2].magic = FLASH_MAP_ENTRY_MAGIC;
+        part_map[2].area.fa_id = FLASH_AREA_IMAGE_SCRATCH;
+        part_map[2].area.fa_device_id = FLASH_DEVICE_ID;
+        part_map[2].area.fa_off = FLASH_AREA_IMAGE_SCRATCH_OFFSET;
+        part_map[2].area.fa_size = FLASH_AREA_IMAGE_SCRATCH_SIZE;
     }
-};
+#if BOOT_IMAGE_COUNT > 1
```

Comment:
for n partitions, perhaps a loop with a macro to assign all this stuff would suffice.

Reply:
I'm not sure how the loop would expand to macros either based on names, or with numbers in the names? The loop is a runtime thing, the names are compile time.

Reply:
If we're just sticking with two images (not N), as was discussed, BOOT_IMAGE_COUNT should become BOOT_DUAL_IMAGE or so.